peps/pep-0751.rst

```diff
@@ -0,0 +1,901 @@
+
+
+Per-file Locking
+''''''''''''''''
+
+*Per-file locking* operates under the premise that one wants to install exact
+same files in any matching environment. As such, the lock file specifies the
+requirements for an environment and then matches those environment requirements
+with the files to install. There can be multiple environments specified in a
+single file, each with their own set of files to install. By specifying the
+exact files to install installers avoid performing a resolution to decide what
+to install.
+
+The motivation for this approach to locking is for those who have controlled
+environments that they work with. For instance, if you have specific, controlled
+development and production environments then you can use per-file locking to
+make sure the **same** files are installed in either environment for everyone.
+
+This is similar to what ``pip freeze`` and pip-tools_
+support, but with more strictness of the exact files as well as incorporating
+support to specify the locked files for multiple environments in the same file.
+
+
+Package Locking
+'''''''''''''''
+
+*Package locking* lists the packages and their versions that *may* apply to any
+environment being installed for. The list of packages and their versions are
+evaluated individually and independently from any other packages and versions
+listed in the file. This allows installation to be linear -- read each package
+and version and make an isolated decision as to whether it should be installed.
+This avoids requiring the installer to perform a *resolution* (i.e.
+determine what to install based on what else is to be installed).
+
+The motivation of this approach comes from
+`PDM lock files <https://frostming.com/en/2024/pdm-lockfile/>`__. By listing the
+potential packages and versions that may be installed, what's installed is
+controlled in a way that's easy to reason about. This also allows for not
+specifying the exact environments that would be supported by the lock file so
+there's more flexibility for what environments are compatible with the lock
+file. This approach supports scenarios like open source projects that want to
+lock what people should use to build the documentation without knowing upfront
+what environments their contributors are working from.
+
+As already mentioned, this approach is supported by PDM_. Poetry_ has
+`shown some interest <https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists/46593/83>`__.
+
+
+Specification
+=============
+
+File Name
+---------
+
+A lock file MUST be named ``pylock.toml`` or match the regular expression
+``f"pylock\.(.+)\.toml"`` if a name is desired or if multiple lock files exist.
+The use of the ``.toml`` file extension is to make syntax highlighting in
+editors easier and to reinforce the fact that the file format is meant to be
+human-readable. The prefix and suffix of a named file MUST be lowercase for easy
+detection and stripping off to find the name, e.g.::
```

Comment:
I think that lock-file names could do with a sentence or two, just saying that lock files can have 'names', what they might be used for, blah -- this paragraph presupposes the idea & dives straight into how to extract the name etc.

```suggestion
A lock file MUST be named :file:`pylock.toml` or match the regular expression
``r"pylock\.(.+)\.toml"`` if a name for the lock file is desired or if multiple lock files exist.
The use of the ``.toml`` file extension is to make syntax highlighting in
editors easier and to reinforce the fact that the file format is meant to be
human-readable. The prefix and suffix of a named file MUST be lowercase for easy
detection and stripping off to find the name, e.g.::
```